<a href="https://colab.research.google.com/github/Namesakenberg/machine_learning/blob/main/crossValidation.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [ ]:
df = pd.read_csv('https://raw.githubusercontent.com/selva86/datasets/master/BostonHousing.csv')

In [ ]:
df.head()

crim    zn  indus  chas    nox     rm   age     dis  rad  tax  ptratio  \
0  0.00632  18.0   2.31     0  0.538  6.575  65.2  4.0900    1  296     15.3   
1  0.02731   0.0   7.07     0  0.469  6.421  78.9  4.9671    2  242     17.8   
2  0.02729   0.0   7.07     0  0.469  7.185  61.1  4.9671    2  242     17.8   
3  0.03237   0.0   2.18     0  0.458  6.998  45.8  6.0622    3  222     18.7   
4  0.06905   0.0   2.18     0  0.458  7.147  54.2  6.0622    3  222     18.7   

        b  lstat  medv  
0  396.90   4.98  24.0  
1  396.90   9.14  21.6  
2  392.83   4.03  34.7  
3  394.63   2.94  33.4  
4  396.90   5.33  36.2

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 506 entries, 0 to 505
Data columns (total 14 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   crim     506 non-null    float64
 1   zn       506 non-null    float64
 2   indus    506 non-null    float64
 3   chas     506 non-null    int64  
 4   nox      506 non-null    float64
 5   rm       506 non-null    float64
 6   age      506 non-null    float64
 7   dis      506 non-null    float64
 8   rad      506 non-null    int64  
 9   tax      506 non-null    int64  
 10  ptratio  506 non-null    float64
 11  b        506 non-null    float64
 12  lstat    506 non-null    float64
 13  medv     506 non-null    float64
dtypes: float64(11), int64(3)
memory usage: 55.5 KB


# the problem with the train test split

In [ ]:
X = df.drop(columns='medv')
y = df['medv']

In [ ]:
from sklearn.model_selection import train_test_split
X_train , X_test , y_train , y_test = train_test_split(X,y,test_size=0.2,random_state=1)

In [ ]:
X_train.shape

(404, 13)

In [ ]:
from sklearn.linear_model import LinearRegression
lr = LinearRegression()
lr.fit(X_train , y_train)
y_pred = lr.predict(X_test)

In [ ]:
from sklearn.metrics import mean_squared_error,r2_score
print("MSE:",mean_squared_error(y_test,y_pred))
print("r2 ",r2_score(y_test,y_pred))

MSE: 23.380836480270336
r2  0.7634174432138461


In [ ]:
# when the random seed is changed the performance changes. hence it is not reliable

## Crossvalidation

Leave one out cross validation

In [ ]:
# in the leave one out method as many no. of models are trained as the number of rows are present in the data
# each row is left for testing once and remaining rows are used for training

In [ ]:
# leave one out
from sklearn.model_selection import cross_val_score,LeaveOneOut
lr = LinearRegression()
loo = LeaveOneOut()
scores = cross_val_score(lr,X,y,cv =loo , scoring='neg_mean_squared_error')
mse_scores = -scores
print("Mean MSE : ",mse_scores.mean())

Mean MSE :  23.72574551947613


In [ ]:
X.shape

(506, 13)

# in this case 506 smaller models are trained
## in this method we DO NOT  get the r2 score because r2 score cannot be calculated for a single row as test data

# K Fold Crossvalidation

In [ ]:
X.shape

(506, 13)

In [ ]:
from sklearn.model_selection import KFold

lr = LinearRegression()
# define k fold
kfold = KFold(n_splits = 10 , shuffle=True , random_state =42)
scores = cross_val_score(lr , X, y , cv=kfold , scoring='r2')

print("r2 scores for each fold : ",scores)
print("Mean r2 scores for k folds : ",scores.mean())

r2 scores for each fold :  [0.75981355 0.60908125 0.76975858 0.71639463 0.61663293 0.79789535
 0.76682601 0.79453027 0.74066667 0.59908146]
Mean r2 scores for k folds :  0.7170680714871446


another one

In [ ]:
url = "https://archive.ics.uci.edu/ml/machine-learning-databases/auto-mpg/auto-mpg.data"
column_names = ['MPG', 'Cylinders', 'Displacement', 'Horsepower', 'Weight', 'Acceleration', 'Model Year', 'Origin']
data = pd.read_csv(url, names=column_names, delim_whitespace=True, na_values='?')


<ipython-input-16-03e20f178e9e>:3: FutureWarning: The 'delim_whitespace' keyword in pd.read_csv is deprecated and will be removed in a future version. Use ``sep='\s+'`` instead
  data = pd.read_csv(url, names=column_names, delim_whitespace=True, na_values='?')


In [ ]:
data.head()

MPG  Cylinders  Displacement  Horsepower  Weight  Acceleration  \
18.0    8      307.0         130.0      3504.0    12.0            70   
15.0    8      350.0         165.0      3693.0    11.5            70   
18.0    8      318.0         150.0      3436.0    11.0            70   
16.0    8      304.0         150.0      3433.0    12.0            70   
17.0    8      302.0         140.0      3449.0    10.5            70   

      Model Year                     Origin  
18.0           1  chevrolet chevelle malibu  
15.0           1          buick skylark 320  
18.0           1         plymouth satellite  
16.0           1              amc rebel sst  
17.0           1                ford torino

In [ ]:
data.shape

(398, 8)

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 398 entries, 18.0 to 31.0
Data columns (total 8 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   MPG           398 non-null    int64  
 1   Cylinders     398 non-null    float64
 2   Displacement  392 non-null    float64
 3   Horsepower    398 non-null    float64
 4   Weight        398 non-null    float64
 5   Acceleration  398 non-null    int64  
 6   Model Year    398 non-null    int64  
 7   Origin        398 non-null    object 
dtypes: float64(4), int64(3), object(1)
memory usage: 28.0+ KB


In [ ]:
data.dropna(inplace = True)

In [ ]:
X = data.drop(columns =['MPG','Origin'])
y = data['MPG']

In [ ]:
X.head()

Cylinders  Displacement  Horsepower  Weight  Acceleration  Model Year
18.0      307.0         130.0      3504.0    12.0            70           1
15.0      350.0         165.0      3693.0    11.5            70           1
18.0      318.0         150.0      3436.0    11.0            70           1
16.0      304.0         150.0      3433.0    12.0            70           1
17.0      302.0         140.0      3449.0    10.5            70           1

In [ ]:
from sklearn.model_selection import train_test_split
X_train , X_test , y_train , y_test = train_test_split(X,y,test_size=0.2,random_state=42)

In [ ]:
X_train.shape


(313, 6)

In [ ]:
lr.fit(X_train ,y_train)
y_pred = lr.predict(X_test)

print("MSE : ",mean_squared_error(y_test,y_pred))
print("r2 : ",r2_score(y_test,y_pred))

MSE :  0.31801123062231107
r2 :  0.8932608319719348


In [ ]:
lr = LinearRegression()
kfold = KFold(n_splits=5 , shuffle=True , random_state=42)
scores = cross_val_score(lr , X_train ,y_train , cv = kfold , scoring='r2')
print("r2 scores for each fold : ",scores)
print("Mean r2 scores for k folds : ",scores.mean())

r2 scores for each fold :  [0.88320931 0.89285836 0.91542477 0.90126258 0.91664586]
Mean r2 scores for k folds :  0.9018801753802919
